File: 2016/src/day09.go
package src

import (
    "fmt"
    "strings"
    "strconv"
)

func decompressed_length(data string, recurse bool) int {
    size := 0
    index := 0
    for index < len(data) {
        if data[index] == '(' {
            marker := index
            for data[marker] != ')' {
                marker++
            }

            split := strings.Split(data[index+1:marker], "x")
            length, _ := strconv.Atoi(split[0])
            times, _ := strconv.Atoi(split[1])
            index = marker + length

            if recurse {
                size += decompressed_length(data[marker+1:marker+1+length], true) * times
            } else {
                size += length * times
            }
        } else {
            size++
        }

        index++
    }

    return size
}

func Solve09(input string) string {
    var ans1 int
    var ans2 int

    ans1 = decompressed_length(input, false)
    ans2 = decompressed_length(input, true)

    return fmt.Sprintf("%d, %d", ans1, ans2)
}

File: 2017/src/day09.js
export default function solve09(input) {
    let ans1 = 0;
    let ans2 = 0;

    function parseGroup(index, depth) {
        ans1 += depth;
        index = parseGroupContent(index + 1, depth + 1);
        return index + 1;
    }

    function parseGarbage(index) {
        ans2--;
        while (input[index] != ">") {
            if (input[index] == "!") {
                index++;
                ans2--;
            }
            index++;
            ans2++;
        }

        return index + 1;
    }

    function parseGroupContent(index, depth) {
        while (input[index] != "}") {
            if (input[index] == ",") {
                index++;
            }

            if (input[index] == "{") {
                index = parseGroup(index, depth);
            } else if (input[index] == "<") {
                index = parseGarbage(index);
            } else {
                console.log(index);
            }
        }

        return index;
    }

    parseGroup(0, 1);

    return [ans1, ans2];
}
